lib/inventory/controller.go

```diff
@@ -208,6 +284,116 @@ func (c *Controller) handleControlStream(handle *upstreamHandle) {
 	}
 }
 
+func (c *Controller) heartbeatInstanceState(handle *upstreamHandle, now time.Time) error {
+	tracker := &handle.stateTracker
+	tracker.MU.Lock()
+	defer tracker.MU.Unlock()
+
+	// withoutLock is used to perform backend I/O outside of the tracker lock. use of this
+	// helper rather than a manual lock/unlock is useful if we ever have a panic in backend logic,
+	// since it preserves the original normal panic rather than causing the less helpful and unrecoverable
+	// 'unlock of unlocked mutex' we would otherwise experience.
+	withoutLock := func(fn func()) {
+		tracker.MU.Unlock()
+		defer tracker.MU.Lock()
+		fn()
+	}
+
+	if tracker.LastHeartbeat == nil {
+		// if we don't know the current heartbeat value, either this is our first time heartbeating this
+		// instance, or we had to reset due to a concurrent update.
+
+		var (
+			lastHB    types.Instance
+			lastRawHB []byte
+			err       error
+		)
+		// perform backend I/O outside of lock
+		withoutLock(func() {
+			lastHB, lastRawHB, err = c.auth.GetRawInstance(c.closeContext, handle.Hello().ServerID)
+		})
+		if err != nil && !trace.IsNotFound(err) {
+			log.Warnf("Failed to load resource for instance %q: %v", handle.Hello().ServerID, err)
+			if !tracker.retryHeartbeat {
+				// suppress failure and retry exactly once
+				tracker.retryHeartbeat = true
+				return nil
+			}
+			// we've already retried, problem is less likely to be transient. close the stream.
+			return trace.Wrap(err)
+		}
+		if err == nil {
+			tracker.LastHeartbeat = lastHB
+			tracker.lastRawHeartbeat = lastRawHB
```

Comment:
What's the difference between a heartbeat and a raw heartbeat?

Reply:
`raw` is the original encoded version of the resource, used for compare and swap.  Will update comments in this function to better explain what's going on with it.